plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcRecordSetProvider.java

```diff
@@ -44,6 +48,31 @@ public JdbcRecordSetProvider(JdbcClient jdbcClient, @ForRecordCursor ExecutorSer
         this.executor = requireNonNull(executor, "executor is null");
     }
 
+    @Override
+    public RecordSet getRecordSet(ConnectorTransactionHandle transaction, ConnectorSession session, ConnectorSplit split, ConnectorTableHandle table, List<? extends ColumnHandle> columns, DynamicFilter dynamicFilter)
+    {
+        if (isEnableDynamicFiltering(session)) {
+            long timeoutMillis = getDynamicFilteringWaitTimeout(session).toMillis();
+
+            // no time for further narrow down or done already
+            if (timeoutMillis == 0 || !dynamicFilter.isAwaitable()) {
+                JdbcTableHandle handle = ((JdbcTableHandle) table).withConstraint(dynamicFilter.getCurrentPredicate());
+                return getRecordSet(transaction, session, split, handle, columns);
+            }
+
+            try {
+                dynamicFilter.isBlocked().get(timeoutMillis, MILLISECONDS);
```

Comment:
This will block an execution thread in the engine and prevent other splits from running, so we need to avoid this somehow.
You can apply dynamic filtering in JdbcSplitManager instead and use `ConnectorSplitSource#getNextBatch` to provide splits with DF using a future there. Changes to ConnectorRecordSetProvider SPI won't be necessary then.